In [303]:
import pandas as pd

In [304]:
df= pd.read_csv(r'C:\Users\maria\Documents\FormacionDataAnalytics\ejercicios\data\online_sales_limpiar.csv', index_col=0)
df.head()

invoice_no stock_code description  quantity         invoice_date  \
0      221958   SKU_1964   WHITE MUG        38  2020-enero-01 00:00   
1      771155   SKU_1241   WHITE MUG        18  2020-enero-01 01:00   
2      231932   SKU_1501  HEADPHONES        49  2020-enero-01 02:00   
3      465838   SKU_1760   DESK LAMP        14  2020-enero-01 03:00   
4      359178   SKU_1386   USB CABLE       -30  2020-enero-01 04:00   

  unit_price  customer_id         country  discount payment_method  \
0       1,71      37039.0       AUSTRALIA  0.470000  Bank Transfer   
1      41,25      19144.0           SPAIN  0.190000        paypall   
2      29,11      50472.0         GERMANY  0.350000  Bank Transfer   
3      76,68      96586.0     NETHERLANDS  0.140000        paypall   
4     -68,11          NaN  UNITED KINGDOM  1.501433  Bank Transfer   

  shipping_cost     category sales_channel  return_status shipment_provider  \
0         10,79      Apparel      IN-STORE              0               UPS   
1          9,51  Electronics        ONLINE              0               UPS   
2         23,03  Electronics        ONLINE              1               UPS   
3         11,08  Accessories        ONLINE              0        Royal Mail   
4           NaN  Electronics      IN-STORE              0             FedEx   

  warehouse_location order_priority  InvoiceNO  
0             LONDON         MEDIUM     221958  
1               ROME         MEDIUM     771155  
2             BERLIN           HIGH     231932  
3               ROME            LOW     465838  
4                NaN         MEDIUM     359178

In [305]:
df.columns

Index(['invoice_no', 'stock_code', 'description', 'quantity', 'invoice_date',
       'unit_price', 'customer_id', 'country', 'discount', 'payment_method',
       'shipping_cost', 'category', 'sales_channel', 'return_status',
       'shipment_provider', 'warehouse_location', 'order_priority',
       'InvoiceNO'],
      dtype='object')

In [306]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 36340 entries, 0 to 36339
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   invoice_no          36340 non-null  int64  
 1   stock_code          36340 non-null  object 
 2   description         36340 non-null  object 
 3   quantity            36340 non-null  int64  
 4   invoice_date        36340 non-null  object 
 5   unit_price          36340 non-null  object 
 6   customer_id         32730 non-null  float64
 7   country             36340 non-null  object 
 8   discount            36340 non-null  float64
 9   payment_method      36340 non-null  object 
 10  shipping_cost       34576 non-null  object 
 11  category            36340 non-null  object 
 12  sales_channel       36340 non-null  object 
 13  return_status       36340 non-null  int64  
 14  shipment_provider   36340 non-null  object 
 15  warehouse_location  33820 non-null  object 
 16  order_pri

```python
<class 'pandas.core.frame.DataFrame'>
Index: 36340 entries, 0 to 36339
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   invoice_no          36340 non-null  int64  
 1   stock_code          36340 non-null  object 
 2   description         36340 non-null  object 
 3   quantity            36340 non-null  int64  
 4   invoice_date        36340 non-null  object 
 5   unit_price          36340 non-null  object `Hay que transformalo a numerico`
 6   customer_id         32730 non-null  float64
 7   country             36340 non-null  object 
 8   discount            36340 non-null  float64
 9   payment_method      36340 non-null  object 
 10  shipping_cost       34576 non-null  object 
 11  category            36340 non-null  object 
 12  sales_channel       36340 non-null  object 
 13  return_status       36340 non-null  int64  
 14  shipment_provider   36340 non-null  object 
 15  warehouse_location  33820 non-null  object 
 16  order_priority      36340 non-null  object 
 17  InvoiceNO           36340 non-null  int64  
dtypes: float64(2), int64(4), object(12)
memory usage: 5.3+ MB


In [307]:
df.isnull().sum()

invoice_no               0
stock_code               0
description              0
quantity                 0
invoice_date             0
unit_price               0
customer_id           3610
country                  0
discount                 0
payment_method           0
shipping_cost         1764
category                 0
sales_channel            0
return_status            0
shipment_provider        0
warehouse_location    2520
order_priority           0
InvoiceNO                0
dtype: int64

In [308]:
df.duplicated().sum()

0

In [309]:
df.describe().T

count           mean            std       min        25%  \
invoice_no     36340.0  551662.412631  260560.241463  100005.0  326151.25   
quantity       36340.0      22.420914      17.816181     -50.0      11.00   
customer_id    32730.0   55014.170975   25957.694626   10001.0   32688.50   
discount       36340.0       0.275637       0.227978       0.0       0.13   
return_status  36340.0       0.098597       0.298124       0.0       0.00   
InvoiceNO      36340.0  551662.412631  260560.241463  100005.0  326151.25   

                     50%        75%            max  
invoice_no     554002.00  777560.75  999997.000000  
quantity           23.00      37.00      49.000000  
customer_id     55245.50   77448.50   99998.000000  
discount            0.26       0.38       1.998946  
return_status       0.00       0.00       1.000000  
InvoiceNO      554002.00  777560.75  999997.000000

In [310]:
# Crear un diccionario para almacenar los resultados
value_counts_dict = {}

# Recorrer las columnas categóricas y almacenar los conteos
for col in df.select_dtypes(include='O').columns:
    value_counts_dict[col] = df[col].value_counts()

# Mostrar todos los resultados de manera ordenada
for col, counts in value_counts_dict.items():
    print(f"Columna: {col}")
    print(counts)
    print('----------------------')


Columna: stock_code
stock_code
SKU_1724    57
SKU_1044    55
SKU_1743    55
SKU_1953    55
SKU_1761    54
            ..
SKU_1212    20
SKU_1273    20
SKU_1126    19
SKU_1122    19
SKU_1031    19
Name: count, Length: 1000, dtype: int64
----------------------
Columna: description
description
WALL CLOCK        3378
HEADPHONES        3357
OFFICE CHAIR      3354
WHITE MUG         3351
USB CABLE         3332
BACKPACK          3307
BLUE PEN          3286
DESK LAMP         3274
NOTEBOOK          3262
WIRELESS MOUSE    3220
T-SHIRT           3219
Name: count, dtype: int64
----------------------
Columna: invoice_date
invoice_date
2021-enero-16 08:00        3
2020-junio-19 14:00        3
2023-febrero-16 04:00      3
2022-noviembre-07 23:00    3
2022-junio-01 23:00        3
                          ..
2021-mayo-11 03:00         1
2021-mayo-11 02:00         1
2021-mayo-11 01:00         1
2021-mayo-11 00:00         1
2023-diciembre-31 23:00    1
Name: count, Length: 35064, dtype: int64
-----------

### Eliminar columnas redundantes

In [311]:
df.columns

Index(['invoice_no', 'stock_code', 'description', 'quantity', 'invoice_date',
       'unit_price', 'customer_id', 'country', 'discount', 'payment_method',
       'shipping_cost', 'category', 'sales_channel', 'return_status',
       'shipment_provider', 'warehouse_location', 'order_priority',
       'InvoiceNO'],
      dtype='object')

In [312]:
df[['invoice_no', 'InvoiceNO']]

invoice_no  InvoiceNO
0          221958     221958
1          771155     771155
2          231932     231932
3          465838     465838
4          359178     359178
...           ...        ...
36335      777548     777548
36336      166019     166019
36337      624739     624739
36338      950706     950706
36339      244458     244458

[36340 rows x 2 columns]

In [313]:
df['invoice_no'].equals(df['InvoiceNO'])

True

In [314]:
df.drop('InvoiceNO', axis=1, inplace=True)

### Homogeneizar los nombres de la columnas

In [315]:
df.columns

Index(['invoice_no', 'stock_code', 'description', 'quantity', 'invoice_date',
       'unit_price', 'customer_id', 'country', 'discount', 'payment_method',
       'shipping_cost', 'category', 'sales_channel', 'return_status',
       'shipment_provider', 'warehouse_location', 'order_priority'],
      dtype='object')

In [316]:
nombres_columnas={col:col.title() for col in df.columns}
nombres_columnas

{'invoice_no': 'Invoice_No',
 'stock_code': 'Stock_Code',
 'description': 'Description',
 'quantity': 'Quantity',
 'invoice_date': 'Invoice_Date',
 'unit_price': 'Unit_Price',
 'customer_id': 'Customer_Id',
 'country': 'Country',
 'discount': 'Discount',
 'payment_method': 'Payment_Method',
 'shipping_cost': 'Shipping_Cost',
 'category': 'Category',
 'sales_channel': 'Sales_Channel',
 'return_status': 'Return_Status',
 'shipment_provider': 'Shipment_Provider',
 'warehouse_location': 'Warehouse_Location',
 'order_priority': 'Order_Priority'}

In [317]:
df.rename(columns=nombres_columnas, inplace=True)
df.columns

Index(['Invoice_No', 'Stock_Code', 'Description', 'Quantity', 'Invoice_Date',
       'Unit_Price', 'Customer_Id', 'Country', 'Discount', 'Payment_Method',
       'Shipping_Cost', 'Category', 'Sales_Channel', 'Return_Status',
       'Shipment_Provider', 'Warehouse_Location', 'Order_Priority'],
      dtype='object')

### Homogeneizar datos categóricos

In [318]:
for col in df.select_dtypes(include='O').columns:
  df[col] =df[col].str.lower()

In [319]:
df.head()

Invoice_No Stock_Code Description  Quantity         Invoice_Date  \
0      221958   sku_1964   white mug        38  2020-enero-01 00:00   
1      771155   sku_1241   white mug        18  2020-enero-01 01:00   
2      231932   sku_1501  headphones        49  2020-enero-01 02:00   
3      465838   sku_1760   desk lamp        14  2020-enero-01 03:00   
4      359178   sku_1386   usb cable       -30  2020-enero-01 04:00   

  Unit_Price  Customer_Id         Country  Discount Payment_Method  \
0       1,71      37039.0       australia  0.470000  bank transfer   
1      41,25      19144.0           spain  0.190000        paypall   
2      29,11      50472.0         germany  0.350000  bank transfer   
3      76,68      96586.0     netherlands  0.140000        paypall   
4     -68,11          NaN  united kingdom  1.501433  bank transfer   

  Shipping_Cost     Category Sales_Channel  Return_Status Shipment_Provider  \
0         10,79      apparel      in-store              0               ups   
1          9,51  electronics        online              0               ups   
2         23,03  electronics        online              1               ups   
3         11,08  accessories        online              0        royal mail   
4           NaN  electronics      in-store              0             fedex   

  Warehouse_Location Order_Priority  
0             london         medium  
1               rome         medium  
2             berlin           high  
3               rome            low  
4                NaN         medium

### Cambio tipo de datos

In [320]:
meses = {
    'enero': '01',
    'febrero': '02',
    'marzo': '03',
    'abril': '04',
    'mayo': '05',
    'junio': '06',
    'julio': '07',
    'agosto': '08',
    'septiembre': '09',
    'octubre': '10',
    'noviembre': '11',
    'diciembre': '12'
}


In [321]:
df['Invoice_Date'] = df['Invoice_Date'].replace(meses, regex=True)



In [322]:
df.head()

Invoice_No Stock_Code Description  Quantity      Invoice_Date Unit_Price  \
0      221958   sku_1964   white mug        38  2020-01-01 00:00       1,71   
1      771155   sku_1241   white mug        18  2020-01-01 01:00      41,25   
2      231932   sku_1501  headphones        49  2020-01-01 02:00      29,11   
3      465838   sku_1760   desk lamp        14  2020-01-01 03:00      76,68   
4      359178   sku_1386   usb cable       -30  2020-01-01 04:00     -68,11   

   Customer_Id         Country  Discount Payment_Method Shipping_Cost  \
0      37039.0       australia  0.470000  bank transfer         10,79   
1      19144.0           spain  0.190000        paypall          9,51   
2      50472.0         germany  0.350000  bank transfer         23,03   
3      96586.0     netherlands  0.140000        paypall         11,08   
4          NaN  united kingdom  1.501433  bank transfer           NaN   

      Category Sales_Channel  Return_Status Shipment_Provider  \
0      apparel      in-store              0               ups   
1  electronics        online              0               ups   
2  electronics        online              1               ups   
3  accessories        online              0        royal mail   
4  electronics      in-store              0             fedex   

  Warehouse_Location Order_Priority  
0             london         medium  
1               rome         medium  
2             berlin           high  
3               rome            low  
4                NaN         medium

In [329]:
df['Invoice_Date']=pd.to_datetime(df['Invoice_Date'], format= "%Y-%m-%d %H:%M")

In [324]:
str_float=['Unit_Price', 'Shipping_Cost']
for col in str_float:
   df[col]=df[col].str.replace(',', '.')
  

In [325]:
for col in str_float:
 df[col]=df[col].apply(lambda x:float(x))

In [326]:
diccionario_map={1:'returned', 0:'not returned'}
df['Return_Status']=df['Return_Status'].map(diccionario_map)


In [327]:
df['Year']= df['Invoice_Date'].dt.year
df['Month']=df['Invoice_Date'].dt.month


In [335]:
df['Day_name']=df['Invoice_Date'].dt.day_name()
df.head()

Invoice_No Stock_Code Description  Quantity        Invoice_Date  \
0      221958   sku_1964   white mug        38 2020-01-01 00:00:00   
1      771155   sku_1241   white mug        18 2020-01-01 01:00:00   
2      231932   sku_1501  headphones        49 2020-01-01 02:00:00   
3      465838   sku_1760   desk lamp        14 2020-01-01 03:00:00   
4      359178   sku_1386   usb cable       -30 2020-01-01 04:00:00   

   Unit_Price  Customer_Id         Country  Discount Payment_Method  ...  \
0        1.71      37039.0       australia  0.470000  bank transfer  ...   
1       41.25      19144.0           spain  0.190000        paypall  ...   
2       29.11      50472.0         germany  0.350000  bank transfer  ...   
3       76.68      96586.0     netherlands  0.140000        paypall  ...   
4      -68.11          NaN  united kingdom  1.501433  bank transfer  ...   

   Sales_Channel Return_Status Shipment_Provider Warehouse_Location  \
0       in-store  not returned               ups             london   
1         online  not returned               ups               rome   
2         online      returned               ups             berlin   
3         online  not returned        royal mail               rome   
4       in-store  not returned             fedex                NaN   

  Order_Priority  Year Month   Day_name  Week Quarter  
0         medium  2020     1  Wednesday     1       1  
1         medium  2020     1  Wednesday     1       1  
2           high  2020     1  Wednesday     1       1  
3            low  2020     1  Wednesday     1       1  
4         medium  2020     1  Wednesday     1       1  

[5 rows x 22 columns]

In [332]:
df['Week']=df['Invoice_Date'].dt.isocalendar().week

In [333]:
df['Quarter']=df['Invoice_Date'].dt.quarter

In [338]:
df.to_csv('data/online_sales_limpio.csv')